### ISSUE TITLE
Treeshaking

### ISSUE THREAD

**Issue by akramsarkis** [Geeks-Solutions/nuxt-sections#74]
- [ ] For a big file, check what part of the file is relevant for your component and extract it

### **Acceptance testing**

- [ ] Check on https://pagespeed.web.dev/ the performances mobile and then desktop

**Comment by akramsarkis** [comment]
@mahfouzhseineweev don't forget that this needs to be tested as well on site not using meta sections so please when done to add an archived version of the library on geeks website dev, like this I can test there as well

**Comment by mahfouzhseineweev** [comment]
## Report:

For this issue as it is mainly about removing unused js codes, I have found a way to analyze the bundlers created upon a build in Nuxt js which is support out of box and can be enabled by adding `analyze: true` to the build in nuxt.config.js file.

Note that Nuxt js which uses Webpack, supports tree-shaking out of the box and no configurations needed to enable it.

After running the analyze I got the below file created which has a summary of all bundlers contents and their sizes.

![Annotation 2024-11-01 163805](https://github.com/user-attachments/assets/529994c4-2a26-40e3-95f9-599a11e59d05)

The quill.js and quill-emoji.js are taking the biggest size across the bundlers and it seems the problem of unused codes is coming from these 2 scripts.

I still have to confirm this as I still need to check in details using the sources coverage that is supported in the devtools out of the box and can point to unused codes by there lines.

The updates needed to adjust the performance for unused js codes is not going to be straight forward and it is unknown currently what are the updates that has to be performed.

As mentioned, more debugging is needed to find what codes exactly are causing the big amount of unused js codes and to check if there is a clean way to remove them

**Comment by jfayad** [comment]
> ## Report:
> For this issue as it is mainly about removing unused js codes, I have found a way to analyze the bundlers created upon a build in Nuxt js which is support out of box and can be enabled by adding `analyze: true` to the build in nuxt.config.js file.
> 
> Note that Nuxt js which uses Webpack, supports tree-shaking out of the box and no configurations needed to enable it.
> 
> After running the analyze I got the below file created which has a summary of all bundlers contents and their sizes.
> 
> ![Annotation 2024-11-01 163805](https://private-user-images.githubusercontent.com/101403093/382312132-529994c4-2a26-40e3-95f9-599a11e59d05.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3MTYyMDQsIm5iZiI6MTczMDcxNTkwNCwicGF0aCI6Ii8xMDE0MDMwOTMvMzgyMzEyMTMyLTUyOTk5NGM0LTJhMjYtNDBlMy05NWY5LTU5OWExMWU1OWQwNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEwNFQxMDI1MDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04NDQ3MjJmNTNmOTdjOWM5OGQxZmI5OThjM2IyNWNhOWNjNTNhODMyZDhhNzAyNzgyZTM2MmJjMjA0YjNkZjU3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.K3BE3VNWd9aCeauDwXXlTranPAWezhXCk9IjVdYURo8)
> 
> The quill.js and quill-emoji.js are taking the biggest size across the bundlers and it seems the problem of unused codes is coming from these 2 scripts.
> 
> I still have to confirm this as I still need to check in details using the sources coverage that is supported in the devtools out of the box and can point to unused codes by there lines.
> 
> The updates needed to adjust the performance for unused js codes is not going to be straight forward and it is unknown currently what are the updates that has to be performed.
> 
> As mentioned, more debugging is needed to find what codes exactly are causing the big amount of unused js codes and to check if there is a clean way to remove them

If treeshaking is performed out of the box, why is quill's codebase so big ? It was detected as an opportunity to be optimized so I'm assuming that a lot of its code is unused... are you sure treeshaking is being performed on Quill's js ?

Are you sure the way the quill plugin is added is "standard" and allow's nuxt's tree shaking to be performed as it normally should ?

**Comment by mahfouzhseineweev** [comment]
> are you sure treeshaking is being performed on Quill's js ?

No this is still unknown

> Are you sure the way the quill plugin is added is "standard" and allow's nuxt's tree shaking to be performed as it normally should ?

It is standard and should allow nuxt tree shaking but I am not sure if quill library itself allow this @jfayad

**Comment by jfayad** [comment]
@mahfouzhseineweev what makes you believe that a library can "forbid" tree shaking from taking place ?

**Comment by mahfouzhseineweev** [comment]
@jfayad I am not sure but I have read the below:

Tree shaking relies on the static structure of ES6 modules (using import and export). Tools like Webpack or Rollup analyze the import/export statements to determine which parts of the code are being used.

And because the library we are using is dead and no more maintained, it could be using old ES modules not supported for tree shaking

**Comment by jfayad** [comment]
I don't think there is such a think that "support tree shaking" or not.
The code in charge of Tree shaking will check which part of the code are not being used and will "prune" it.

Now if the code in question is not written in a modular way, the tree shaking process could consider that the whole code is being used (even if only a tiny part of it is useful).

If that's what's happening here, maybe we could push a PR to separate some pieces of codes that way tree shaking will be able to perform its duty ?

**Comment by mahfouzhseineweev** [comment]
With the new performance reports we had on pagespeed, I haven't noticed any leak of unused javascript specific to quill which had a big bad impact previously with nuxt 2 and quill 1.

I need to recheck the performance with the quill sections in use and make sure this is no more an issue.
There are also other unused javascript that has to be checked to see if they can be solved.

This library seems to have a good way of enhancing page performance and should be checked and tested
https://nuxt.com/modules/nuxt-booster

A Timebox on this issue is needed

**Comment by akramsarkis** [comment]
> With the new performance reports we had on pagespeed, I haven't noticed any leak of unused javascript specific to quill which had a big bad impact previously with nuxt 2 and quill 1.
> 
> I need to recheck the performance with the quill sections in use and make sure this is no more an issue. There are also other unused javascript that has to be checked to see if they can be solved.
> 
> This library seems to have a good way of enhancing page performance and should be checked and tested https://nuxt.com/modules/nuxt-booster
> 
> A Timebox on this issue is needed

@jfayad  as you can see hussein needs a time box on it to define exactly the needs,  knowing that he already spent  time on it, I think like 30 40mn. Should we include the tinebox on 3.0.5 or move it to 3.0.6 ?

**Comment by mahfouzhseineweev** [comment]
Indeed, after implementing the Main component refactoring and as there is no improvement in the performance, a Timebox is needed to identify what resources are still getting retrieved by the page when visiting the site as a visitor

**Comment by mahfouzhseineweev** [comment]
### Report:

I have installed and configured the nuxt-booster module, and after performing a build the Tailwind css classes are removed and so the sections are broken.

I tried disabling the config that is responsible of updating the css chunks but still Tailwind classes are removed.
I tried loading the nuxt-booster module before the Tailwind one.

I checked on the module repository issues to know if anyone reported such a problem but I couldn't find any.
I still don't know what exactly is causing the Tailwind classes to be removed.

Also checked the lighthouse performance with that even though this is not a valid test as the site css is broken but I noticed that there is no change in the performance measures we previously had.

I think I should be reporting this problem to the module repo and see if I can get help to unlock the Tailwind removed problem and then identify what other changes are needed to be configured on the project to have better results

**Comment by jfayad** [comment]
agreed @mahfouzhseineweev
